### PR TITLE
Globalization support for HTML ouput

### LINKED ISSUE #15: Right-To-Left support
Hi, I was wondering what would it take to add a new module to wyam to support RTL content. The module needs to add dir="rtl" attribute to all elements that start with RTL characters and align="right" to tables, RTL characters are specified by section D of RFC3454. For more information, please take a look at lunet-io/markdig#239.

It would also be useful if the module adds a css class to these html elements so that they can be styled (most probably their font characteristics) separately.

I am willing to create a PR, but might need some help and guidance.

### PR DESCRIPTION
**Pull request by encrypt0r**
> Note: #73 had gone stale and too much stuff had changed so I was not able to cleanly rebase it. So I am resuming work here.

This is a little bigger than I first thought. Because in lunet-io/markdig#239 I was dealing with a very limited number of element types. But here in Statiq, we are dealing with full html. So I am worried about the performance impact of the module.

After some research, I have found out these things:

1. Browsers don't handle paragraph direction detection very well. [See this talk](https://youtu.be/XgqP0qogg6U?t=1599).
1. [`dir=auto` is still not sufficient](https://twitter.com/fantasai/status/1208199882093162497) for today's browsers. We still need to specify the direction of an html element whenever possible.
1. Most common technique for finding the direction of a block of text is finding the `First strong character`.
1. Tables need `align=right` for rtl text in order to render correctly.
1. We should [prefer HTML markup over CSS](https://www.w3.org/International/questions/qa-bidi-css-markup) for specifying
1. All [block-level elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Block-level_elements#Elements) need to be marked with "rtl" attribute if needed.

## Additional reading
1. https://www.w3.org/International/articles/inline-bidi-markup/uba-basics

Fixes #15

### CODE REVIEW

**Comment by encrypt0r** [comment]
@daveaglick I have a few questions:
1. Where do you think is the best place to put the utility funtions like `IsLeftToRight`, `IsRightToLeft`, `ToUtf32`, `IsBlockElement` and `GetTextDirection`?
1. Part of the code is copied from [MarkDig's Globalization extension](https://github.com/lunet-io/markdig/blob/master/src/Markdig/Extensions/Globalization/GlobalizationExtension.cs) (although I am the original author of that module), how do you think we should deal with the licensing?

**Comment by encrypt0r** [comment]
Hi @daveaglick, are we blocked on anything here? **I understand that you might be very busy in these crazy times**. I just wanted to know if you are waiting on me. Please let me know if I am supposed to do something for this PR to go forward.

**Comment by daveaglick** [comment]
Nope, not blocked at all - this is all great from my first go through it. I’m on vacation this week so I’m hoping to get some time to do a final pull and get this merged in for the next release. Thanks again!

File: src/extensions/Statiq.Html/Globalization.cs [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+        }
+
+        protected async override Task<IEnumerable<Common.IDocument>> ExecuteContextAsync(IExecutionContext context)
+        {
+            List<Common.IDocument> list = new List<Common.IDocument>();
+
+            HtmlParser parser = new HtmlParser();
+            foreach (Common.IDocument input in context.Inputs)
+            {
+                IHtmlDocument htmlDocument = await input.ParseHtmlAsync(context, parser);
```
Comment (daveaglick) [inline src/extensions/Statiq.Html/Globalization.cs:27]: Nice use of the `.ParseHtmlAsync()` extension, glad you found that

File: src/extensions/Statiq.Html/Globalization.cs [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+using System.IO;
+using System.Threading.Tasks;
+using AngleSharp.Dom;
+using AngleSharp.Dom.Html;
+using AngleSharp.Parser.Html;
+using Statiq.Common;
+
+namespace Statiq.Html
+{
+    public class Globalization : Module
```
Comment (daveaglick) [inline src/extensions/Statiq.Html/Globalization.cs:11]: Is there anything about this module that would prevent from processing documents in parallel? If not, we could derive from ParallelModule instead and move the `HtmlParser` up to a class-level static.

Comment (daveaglick) [inline src/extensions/Statiq.Html/Globalization.cs:11]: Can we rename this module to something that follows the [Verb][Noun] convention? Maybe `AddGlobalization`, `AddGlobalizationAttributes`, or `GlobalizeAttributes` (I'm open to other suggestions)?

File: tests/extensions/Statiq.Html.Tests/GlobalizationFixture.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+using System.Threading.Tasks;
+using NUnit.Framework;
+using Shouldly;
+using Statiq.Common;
+using Statiq.Testing;
+
+namespace Statiq.Html.Tests
+{
+    [TestFixture]
+    public class GlobalizationFixture : BaseFixture
```
Comment (daveaglick) [inline tests/extensions/Statiq.Html.Tests/GlobalizationFixture.cs:13]: You're my hero for adding a great set of tests!

**Comment by daveaglick** [comment]
Finally getting a chance to review all this, thanks again and sorry it's taken so long!

In general I really like the design. Implementing the HTML rewriting from a module in Statiq.Html is _exactly_ how I would have done it because that makes it opt-in for anyone that wants to add this functionality. I'll add the module to the default pipelines in Statiq.Web once this gets merged and released with a feature flag so it can be turned on there too.

I just have a few comments, mostly nits and semantics. Otherwise this looks really clean and I’m happy to merge it in.

> Where do you think is the best place to put the utility funtions like IsLeftToRight, IsRightToLeft, ToUtf32, IsBlockElement and GetTextDirection?

Adding them as static helpers in the module as you've done seems like a good approach for now. We can always move them out into a separate static helper class later if it turns out we need that kind of functionality in more than one place.

> Part of the code is copied from MarkDig's Globalization extension (although I am the original author of that module), how do you think we should deal with the licensing?

I appreciate you being mindful of licensing. Markdig looks like BSD 2-Clause which is really similar to MIT. One of those clauses says “Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.”, though I’m not totally certain copying individual blocks of code qualifies as a redistribution. @xoofx any thoughts on how you’d like to see this code from the Markdig module attributed? I’m happy to go with whatever you feel is appropriate.

**Comment by xoofx** [comment]
> @xoofx any thoughts on how you’d like to see this code from the Markdig module attributed?

No need to complicate, if the code copied is from what @encrypt0r did in Markdig, then no need to add a mention to Markdig license here.

**Comment by daveaglick** [comment]
Thanks @xoofx for the confirmation, though actually this makes me question if I'm properly attributing Markdig in general (separate issue, but I'll make sure to verify ASAP).

File: src/extensions/Statiq.Html/Globalization.cs (hunk 1 above)
Comment (encrypt0r) [reply src/extensions/Statiq.Html/Globalization.cs]: Thanks :)

File: src/extensions/Statiq.Html/Globalization.cs (hunk 2 above)
Comment (encrypt0r) [reply src/extensions/Statiq.Html/Globalization.cs]: I renamed it to `AddRtlSupport`. That'll make it clear what it does. And many people (especially those whose first language is not English), might not now what is Globalization. This also follows the [Verb][Noun] convention. What do you think?

**Comment by encrypt0r** [comment]
@daveaglick Thank you very much. I love the code base, it's very well maintained and organized. Let me know if there was anything else needed here

**Comment by daveaglick** [comment]
Thanks again! Hopefully I'll get a release out in the next couple days.

**Pull request merged**